scalyr_agent/builtin_monitors/syslog_monitor.py

```diff
@@ -1084,28 +1111,25 @@ def handle(self):
 
         try:
             if self.__request_processing_executor:
-                # Worker is responsible for processing the data read from one request.
-                # Using the queue ensures the data is processed in the order it was read.
-                DONE = "DONE"
-                work_queue = queue.Queue()
-                def worker(queue, is_shutdown):
-                    data = queue.get(block=True)
-                    while data != DONE:
-                        if is_shutdown():
-                            global_log.info("ThreadPool shutting down, skipping further request processing.")
-                            break
-                        self.__request_data_process(syslog_parser, data)
-                        data = queue.get(block=True)
-
-                self.__request_processing_executor.submit(worker, work_queue, lambda: self.__request_processing_executor._shutdown)
-
+                # Using last_future to ensure the data from one requests is processed in the same order as it was received
+                last_future = None
                 for data in self.__request_stream_read(syslog_request, self.server.is_running):
-                    work_queue.put(data)
-
-                work_queue.put(DONE)
+                    try:
+                        last_future = self.__request_processing_executor.submit(self.__worker, syslog_parser.process, data, last_future)
+                    except RuntimeError:
+                        # There's no way of telling what happened to the executor, it does not have a specific exception class for this case.
+                        global_log.warn("Cannot submit futher data for processing. The server is probably shutting down.")
+                        break
             else:
-                for data in self.__request_stream_read(syslog_request, self.server.is_running):
-                    self.__request_data_process(syslog_parser, data)
+                try:
+                    for data in self.__request_stream_read(syslog_request, self.server.is_running):
+                        syslog_parser.process(data)
+                except Exception as e:
+                    global_log.warning(
+                        "Error processing request: %s\n\t%s",
+                        six.text_type(e),
+                        traceback.format_exc(),
+                    )
```

Comment:
The try/except should probably be inside the for loop (one bad iteration shouldn't trash the whole batch)

Reply:
I wasn't sure about that. But it wouldn't hurt to try to process even partial data.